### DISCUSSION TITLE
dependency source specification in FastAPI's Dependency Injection system

### DISCUSSION THREAD

**Discussion by markkorput** [fastapi/fastapi#3230] [Questions]
When using the Dependency Injection system of FastAPI to get for example a DB interface instance in your request handler, all examples and [documentation](https://fastapi.tiangolo.com/tutorial/sql-databases/) suggest doing something like `db: Session = Depends(get_db)` where `get_db` is a function that returns the Session instance.

However, this way the request handler code is **directly coupled to the instantiation of the `Session` instance** and thus -directly or indirectly- to the configuration of the execution environment. I would like to achieve stronger decoupling, where the app-code can provide dependencies and the request handlers only specify which dependencies they need, without being involved with where these dependencies come from.

This is the kind of behaviour I've come to expect working with DI in Angular (and other environments), where Components and Services specify their dependencies and Modules provide them, so I'm a bit surprised with the way DI is implemented in FastAPI. 

I've also seen python libraries that seem to behave more like the way I would expect (like [injector](https://github.com/alecthomas/injector)), but I did see a lot of people express enthusiasm with the FastAPI DI approach. So I guess my questions are twofold;

1. Is what I want to do possible with FastAPI? Is there a (clean) way to receive a DB `Session` instance in my request handler, which is being instantiated elsewhere (somewhere higher up in de dependecy chain) without my handler having to specify _how_ to instantiate?
2. Is there a specific reason what makes the suggested FastAPI so powerful that I might be missing? (shouldn't I be trying to do what I am asking in 1.?)

**Comment by markkorput** [comment]
Alright, so to answer the first part of my own question, I did something like this;

```python
# app.py

@app.middleware('http')
async def add_db_session_to_request(request: Request, call_next: Callable) -> Any:
    request.state.sessionmaker = sessionmaker(create_engine(DATABASE_URL))
    return await call_next(request)
```

```python
# service.py

async def db_session(request: Request) -> Session:
    async with request.state.sessionmaker() as session:
        yield session

@app.get("/posts", response_model=PostsSchema)
def get_posts(session: Session = Depends(db_session)
    return session.query(Post).all()
```

I'm still curious to hear if this is a smart approach or just not "the FastAPI way" to go about doing things. I like that it hides the DB configuration details from the request handlers, but it feels a bit hacky an boilerplaty. Would love it if FastAPI supported something like this out of the box,  kinda like -for example- how pytest injects fixtures values.

Feedback welcome.

**Reply by tiangolo** [reply]
Thanks for the interest!

I'm not sure I'm being able to understand what is the use case that you are trying to solve and that you can't solve with the current system. Maybe you could explain a bit more what you want to achieve (independent of how it is implemented) and that can help me and others guide you better.

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.

**Comment by r-bar** [comment]
Something that is a little buried in the docs is a couple of very important facts:
1. Dependencies can have access to other dependencies (and anything that can be injected into the routes like query params, the request object, the response object, etc.)
2. Dependencies are keyed by their `id()` by default

This results in a strong affinity for a "default set of dependencies" that get used by the application since they get refered to by referencing their "default implementation". The primary way of overriding them is via setting `FastAPI().dependency_overrides`. I generally only use the overrides in tests. Any application application level polymorphism is handled in the dependency chain.

Let's take the example of what you seem to be trying to do in particular: have a configurable database url and create a session object using that database connection. In my own app I have something like this:

```python
from fastapi import Depends
from pydantic import BaseSettings
from sqlalchemy import create_engine
from sqlalchemy.orm import Session, sessionmaker as sa_sessionmaker

class Config(BaseSettings):
    database_url: str

async def sessionmaker(config=Depends(Config)):
    return sa_sessionmaker(create_engine(config.database_url))

async def db_session(sessionmaker=Depends(sessionmaker)):
    async with sessionmaker() as session:
        yield session

@app.get('/')
async def index(session: Session = Depends(db_session)):
    return await session.query(Post).all()
```
> Note: In a real application I would probably be doing some caching of the engine / sessionmaker so they are not recreated every request

I have not used Angular, so let me know if I missed the mark on trying to solve the example you gave.

**Reply by tiangolo** [reply]
Thanks for the help here @r-bar ! 🙇
